File: TableOfContentsSample/Classes/LayerInfoCell.m
// Copyright 2012 ESRI
//
// All rights reserved under the copyright laws of the United States
// and applicable international laws, treaties, and conventions.
//
// You may freely redistribute and use this sample code, with or
// without modification, provided you include the original copyright
// notice and use restrictions.
//
// See the use restrictions at http://help.arcgis.com/en/sdk/10.0/usageRestrictions.htm
//

#import "LayerInfoCell.h"

#define IMG_HEIGHT_WIDTH 20
#define CHECKBOX_HEIGHT_WIDTH 25
#define CELL_HEIGHT 44
#define SCREEN_WIDTH 320
#define LEVEL_INDENT 32
#define YOFFSET 12
#define XOFFSET 6

@interface LayerInfoCell (Private)

- (UILabel *)labelWithPrimaryColor:(UIColor *)primaryColor 
						selectedColor:(UIColor *)selectedColor 
							 fontSize:(CGFloat)fontSize 
								 bold:(BOOL)bold;
- (IBAction)checkBoxClicked;

@end

@interface CheckBox : UIButton {
	BOOL _isChecked;
}

@property (nonatomic,readonly) BOOL isChecked;

- (id)initWithState:(BOOL)state;
- (void)changeCheckBox;

@end

@implementation LayerInfoCell

@synthesize valueLabel = _valueLabel, visibilitySwitch = _visibilitySwitch, arrowImage = _arrowImage;
@synthesize level = _level, expanded = _expanded, canChangeVisibility = _canChangeVisibility, layerInfoCellDelegate = _layerInfoCellDelegate;

- (id)initWithStyle:(UITableViewCellStyle)style 
	reuseIdentifier:(NSString *)reuseIdentifier 
			  level:(NSUInteger)level 
canChangeVisibility:(BOOL)canChangeVisibility
         visibility:(BOOL)visibility
		   expanded:(BOOL)expanded;  {
    
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
		self.level = level;       
        self.canChangeVisibility = canChangeVisibility;
		self.expanded = expanded;
		
		UIView *content = self.contentView;
		
		self.valueLabel = 
        [self labelWithPrimaryColor:[UIColor blackColor] 
                         selectedColor:[UIColor whiteColor] 
                              fontSize:16.0 bold:YES];
		self.valueLabel.textAlignment = UITextAlignmentLeft;
		[content addSubview:self.valueLabel];
        
        if(_canChangeVisibility)
        {
            self.visibilitySwitch = [[CheckBox alloc] initWithState:visibility];
            [self.visibilitySwitch setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
            [self.visibilitySwitch setTitle:@"" forState:UIControlStateNormal];   
            [self.visibilitySwitch addTarget:self action:@selector(visibilityChanged) forControlEvents:UIControlEventTouchUpInside];
            [content addSubview:self.visibilitySwitch];
        }             
    
		self.arrowImage = [[UIImageView alloc] initWithImage:[UIImage imageNamed:self.expanded ? @"CircleArrowDown_sml" : @"CircleArrowRight_sml"]];
		[content addSubview:self.arrowImage];
    }
    return self;
}


#pragma mark -
#pragma mark Memory Management

- (void)dealloc {
	self.layerInfoCellDelegate = nil;
}

#pragma mark -
#pragma mark Other overrides

- (void)layoutSubviews {
    [super layoutSubviews];
    CGRect contentRect = self.contentView.bounds;
	
    if (!self.editing) {
		
		// get the X pixel spot
        CGFloat boundsX = contentRect.origin.x;
		CGRect frame;
        
        CGRect imgFrame = CGRectMake(((boundsX + self.level + 1) * LEVEL_INDENT) - (IMG_HEIGHT_WIDTH + XOFFSET), 
							  YOFFSET, 
							  IMG_HEIGHT_WIDTH, 
							  IMG_HEIGHT_WIDTH);
		self.arrowImage.frame = imgFrame;
        
        if(self.canChangeVisibility)
        {           
            CGRect checkBoxFrame = CGRectMake(((boundsX + self.level + 1) * LEVEL_INDENT), 
                                       (CELL_HEIGHT - CHECKBOX_HEIGHT_WIDTH)/2, 
                                       CHECKBOX_HEIGHT_WIDTH,
                                       CHECKBOX_HEIGHT_WIDTH);
            self.visibilitySwitch.frame = checkBoxFrame;
            
            frame = CGRectMake((boundsX + self.level + 1) * LEVEL_INDENT + 30, 
                               0, 
                               SCREEN_WIDTH - (self.level * LEVEL_INDENT), 
                               CELL_HEIGHT);
            self.valueLabel.frame = frame;           
            
        }
		
		else
        {
            frame = CGRectMake((boundsX + self.level + 1) * LEVEL_INDENT, 
                               0, 
                               SCREEN_WIDTH - (self.level * LEVEL_INDENT), 
                               CELL_HEIGHT);
            self.valueLabel.frame = frame;
        }
		
	}
}

#pragma mark -
#pragma mark Private category

- (UILabel *)labelWithPrimaryColor:(UIColor *)primaryColor 
						selectedColor:(UIColor *)selectedColor 
							 fontSize:(CGFloat)fontSize 
								 bold:(BOOL)bold {
    UIFont *font;
    if (bold) {
        font = [UIFont boldSystemFontOfSize:fontSize];
    } else {
        font = [UIFont systemFontOfSize:fontSize];
    }
	
	UILabel *newLabel = [[UILabel alloc] initWithFrame:CGRectZero];
	newLabel.backgroundColor = [UIColor whiteColor];
	newLabel.opaque = YES;
	newLabel.textColor = primaryColor;
	newLabel.highlightedTextColor = selectedColor;
	newLabel.font = font;
	newLabel.numberOfLines = 0;
	
	return newLabel;
}

- (IBAction)visibilityChanged {       
    [self.visibilitySwitch changeCheckBox];
    [self.layerInfoCellDelegate layerVisibilityChanged:self.visibilitySwitch.isChecked forCell:self];
}

@end



@implementation CheckBox

@synthesize isChecked = _isChecked;

- (id)initWithState:(BOOL)state {
    if (self = [super init]) {
        // Initialization code
		
		//self.frame =frame;
		self.contentHorizontalAlignment  = UIControlContentHorizontalAlignmentLeft;       

        if(state) {
            [self setImage:[UIImage imageNamed:@"checkbox_checked.png"] forState:UIControlStateNormal];	    
        }
        else {
            [self setImage:[UIImage imageNamed:@"checkbox_unchecked.png"] forState:UIControlStateNormal];	
        }
        _isChecked = state;			
	}
    return self;
}

- (void)changeCheckBox {
    _isChecked = !_isChecked;
    
    if(_isChecked) {		
		[self setImage:[UIImage imageNamed:@"checkbox_checked.png"] forState:UIControlStateNormal];			
	} else {
		[self setImage:[UIImage imageNamed:@"checkbox_unchecked.png"] forState:UIControlStateNormal];        
	}       
}





@end

File: TableOfContentsSample/swift/TableOfContents/TOC/Cells/LayerInfoCell.swift
//
// Copyright 2015 ESRI
//
// All rights reserved under the copyright laws of the United States
// and applicable international laws, treaties, and conventions.
//
// You may freely redistribute and use this sample code, with or
// without modification, provided you include the original copyright
// notice and use restrictions.
//
// See the use restrictions at http://help.arcgis.com/en/sdk/10.0/usageRestrictions.htm
//

import UIKit

let LEVEL_INDENT:CGFloat = 15

class CheckBox:UIButton {
    
    var isChecked:Bool = false {
        didSet {
            self.updateCheckBox()
        }
    }
    
    func changeCheckBox() {
        self.isChecked = !self.isChecked
    }
    
    func updateCheckBox() {
        if self.isChecked {
            self.setImage(UIImage(named: "checkbox_checked.png"), forState:.Normal)
        } else {
            self.setImage(UIImage(named: "checkbox_unchecked.png"), forState:.Normal)
        }
    }
}

protocol LayerInfoCellDelegate:class {
    func layerInfoCell(layerInfoCell:LayerInfoCell, didChangeVisibility visibility:Bool)
}

class LayerInfoCell: UITableViewCell {
    
    @IBOutlet weak var valueLabel:UILabel!
    @IBOutlet weak var visibilitySwitch:CheckBox!
    @IBOutlet weak var arrowImage:UIImageView!
    
    @IBOutlet weak var arrowImageLeadingConstraint: NSLayoutConstraint!
    @IBOutlet weak var visibilitySwitchLeadingConstraint: NSLayoutConstraint!
    
    
    var level:Int!
    var expanded:Bool = false {
        //update the arrow image based on the state of the expansion
        didSet {
            if self.arrowImage != nil {
                self.arrowImage.image = UIImage(named: self.expanded ? "CircleArrowDown_sml" : "CircleArrowRight_sml")
            }
        }
    }
    var canChangeVisibility:Bool = false {
        //show checkbox if the layer can be hidden
        didSet {
            canChangeVisibility ? self.showVisibilitySwitch() : hideVisibilitySwitch()
        }
    }
    var visibility:Bool = false {
        //set the state of the checkbox
        didSet {
            self.visibilitySwitch?.isChecked = visibility
        }
    }
    
    weak var layerInfoCellDelegate:LayerInfoCellDelegate?
    
    //called when the user checks or unchecks the visibility switch
    @IBAction func visibilityChanged() {
        self.visibilitySwitch.changeCheckBox()
        self.layerInfoCellDelegate?.layerInfoCell(self, didChangeVisibility: self.visibilitySwitch.isChecked)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        //set the indentation based on the level
        self.arrowImageLeadingConstraint.constant = CGFloat(self.level + 1) * LEVEL_INDENT
    }
    
    //MARK: - hide/show switch
    
    func hideVisibilitySwitch() {
        self.visibilitySwitch.hidden = true
        self.visibilitySwitchLeadingConstraint.constant = -self.visibilitySwitch.bounds.size.width
    }
    
    func showVisibilitySwitch() {
        self.visibilitySwitch.hidden = false
        self.visibilitySwitchLeadingConstraint.constant = 8
    }
}
